poollet/machinepoollet/mem/generic.go

```diff
@@ -0,0 +1,151 @@
+// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and IronCore contributors
+// SPDX-License-Identifier: Apache-2.0
+
+package mem
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/gogo/protobuf/proto"
+	computev1alpha1 "github.com/ironcore-dev/ironcore/api/compute/v1alpha1"
+	"github.com/ironcore-dev/ironcore/iri/apis/machine"
+	iri "github.com/ironcore-dev/ironcore/iri/apis/machine/v1alpha1"
+	"github.com/ironcore-dev/ironcore/poollet/machinepoollet/api/v1alpha1"
+	"golang.org/x/exp/maps"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type Generic struct {
+	record.EventRecorder
+	client.Client
+	mu sync.RWMutex
+
+	sync   bool
+	synced chan struct{}
+
+	machineEventsByID map[string]*iri.MachineEvents
+
+	machineRuntime machine.RuntimeService
+
+	relistPeriod time.Duration
+}
+
+func isNewEventsPresent(oldMachineEventsByID map[string]*iri.MachineEvents, machineEvent *iri.MachineEvents, machineID string) bool {
+	oldMachineEvent, ok := oldMachineEventsByID[machineID]
+	if !ok {
+		return true
+	}
+	return !proto.Equal(machineEvent, oldMachineEvent)
+}
+
+func (g *Generic) relist(ctx context.Context, log logr.Logger) error {
+	log.V(1).Info("Relisting machine cluster events")
+	toEventFilterTime := time.Now()
+	fromEventFilterTime := toEventFilterTime.Add(-1 * g.relistPeriod)
+	res, err := g.machineRuntime.ListEvents(ctx, &iri.ListEventsRequest{
+		Filter: &iri.EventFilter{EventsFromTime: fromEventFilterTime.Unix(), EventsToTime: toEventFilterTime.Add(-1 * g.relistPeriod).Unix()},
+	})
+	if err != nil {
+		return fmt.Errorf("error listing machine cluster events: %w", err)
+	}
+
+	g.mu.Lock()
+	defer g.mu.Unlock()
+
+	oldMachineEventsByID := maps.Clone(g.machineEventsByID)
+
+	maps.Clear(g.machineEventsByID)
+
+	var shouldPublishEvents bool
+	for _, machineEvent := range res.MachineEvents {
+		if machine, err := g.getMachine(ctx, machineEvent.InvolvedObjectMeta.GetLabels()); err == nil {
+			shouldPublishEvents = isNewEventsPresent(oldMachineEventsByID, machineEvent, string(machine.GetUID()))
+			if shouldPublishEvents {
+				for _, event := range machineEvent.Events {
+					g.Eventf(machine, event.Spec.Type, event.Spec.Reason, event.Spec.Message)
+				}
+			}
+			g.machineEventsByID[string(machine.GetUID())] = machineEvent
+		}
+	}
+
+	if !g.sync {
+		g.sync = true
+		close(g.synced)
+	}
+
+	return nil
```

Comment:
We might want to do this in a simpler fashion without storing events. What do you think about the following?
- Introduce a variable something like `lastFetched` and set it on startup to the current time
- While running always fetch events from `lastFetched` until `Now`

Reply:
Done, simplified logic without storing events